Values can have different data types

claire-lovisa opened the pull request:
# Context

[OP-3584]

To be tested with https://github.com/lblod/app-organization-portal/pull/537

Values fetched from wegwijs were always stored as string, but they can actually have other types as well (here date and uri). This PR adds the correct data type when updating the db.

## Side quests

1. I also bumped some dev dependencies, something was off and was preventing the service to start
2. I updated the tests accordingly as well

# How to test

Please refer to https://github.com/lblod/app-organization-portal/pull/537 and its changelog for test instructions :)

claire-lovisa committed:
Fix status being a URI and start/modify being dates

claire-lovisa committed:
Remove extra spaces

claire-lovisa committed:
Bump dev deps

claire-lovisa committed:
Values can be undefined

claire-lovisa committed:
Update Woodpecker CI config files

claire-lovisa committed:
Merge pull request #10 from lblod/chore/woodpecker-config-updates

Update Woodpecker CI config files

mirdono reviewed (commented):
**Suggestion:** Imo passing around a string argument and checking it in a switch statement is bad practice. This kind of code tends to be fragile and annoying to extend/maintain over time. A typo in a string is easily made and overlooked, possibly leading to annoying bugs to find. At least, I would define the allowed strings in an `enum` or something to simplify passing the right strings. 

Building on such an enum, instead of strings you might just pass the name of the necessary escape function and use `apply` instead of the `switch` statement. Something along the following of the following snippet. Note, in the snippet I also changed to order of the parameters to make the `valueEscapeFunction` optional, and gave it a default value. (At this point, I am not sure anymore whether the `enum` has added value compared to just specifying the function names.)

``` javascript
const dataValueType = {
  STRING: sparqlEscapeString,
  DATE: sparqlEscapeDate,
  URI: sparqlEscapeUri,
}

export const buildUpdateQuery = (
 subject,
  type,
  predicate
  graphValues, // NOTE: order of these parameters changed
  value,
  valueEscapeFunction = sparqlEscapeString,
) => {
  const escapedSubject = sparqlEscapeUri(subject);
  const escapedPredicate = sparqlEscapeUri(predicate);

  let escapedValue = value ? valueEscapeFunction.apply(null, [value]) : value;
  
  const deleteQuery = ...
  const insertQeury = ...
  return `${deleteQuery} ${insertQuery}`;
}

...

const updateKboOrganization = async (kboOrgUri, kboFields) => {
  const updateRechtsvormQuery = buildUpdateQuery(
    kboOrgUri,
    "http://mu.semte.ch/vocabularies/ext/KboOrganisatie",
    "http://mu.semte.ch/vocabularies/ext/rechtsvorm",
    graphValues,
    kboFields.rechtsvorm,
    // NOTE: not need to specify dataValueType.STRING
  );
  
  ...
  
  const updateModifiedQuery = buildUpdateQuery(
    kboOrgUri,
    "http://mu.semte.ch/vocabularies/ext/KboOrganisatie",
    "http://purl.org/dc/terms/modified",
    graphValues,
    kboFields.changeTime,
    dataValueType.DATE, // NOTE: replaces the 'date' string
  );
  
  ...
}

...
```

Note, I could not get this properly working with the defined tests. Possibly the mocked escape functions cause some issues (in my setup). Running a  healing job with this version does seem to give the expected results. But I did not inspect them in depth. 


Another possibility is to let the calling functions call the right escape function directly on the passed `value`. This allows to remove escaping `value` from `buildUpdateQuery` altogether. A disadvantage here is that each of the calling functions do have to properly deal with `undefined` values (which is now centralised in the `buildUpdateQuery`.) One way to deal with this might to define wrapper functions around the actual escape functions, with right naming this can be transparant:

``` javascript
import { sparqlEscapeString as muSparqlEscapeString } from "mu"; // NOTE: changed the name

function sparqlEscapeString(value) {
  return value ? muSparqlEscapeString(value) : undefined;
}

...

export const buildUpdateQuery = (
  subject,
  type,
  predicate,
  escapedValue, // <- renamed the original `value`
  graphValues,
) => {
  const escapedSubject = sparqlEscapeUri(subject);
  const escapedPredicate = sparqlEscapeUri(predicate);
  
  // NOTE: no need to escape `escapedValue` here
  
  const deleteQuery = ..

  const insertQuery = escapedValue
    ? ...
    : "";

  return `${deleteQuery} ${insertQuery}`;
};

const updateKboAddress = async (addressUri, formattedAddress) => {
  const query = buildUpdateQuery(
    addressUri,
    "http://www.w3.org/ns/locn#Address",
    "http://www.w3.org/ns/locn#fullAddress",
    sparqlEscapeString(formattedAddress), // <- replaces `'string', formattedAddress`
    graphValues
  );

  await update(query);
};

...
```

claire-lovisa committed:
Use an enum for data type instead of strings

claire-lovisa commented:
@mirdono thanks for the detailed feedback. You're completely right, strings is far from the best approach. For your two solution, I went for the first one with enums because I feel like it's the one that is the easier to understand when reading the code. The wrappers of solution two are I think easy to miss.

I implemented your suggestion and adjusted the test as well.

mirdono committed:
chore: removed trailing blank space

mirdono reviewed (approved):
Looks good. I removed some additional blank spaces that caught my eye.
